pkg/gitops/profile.go

```diff
@@ -0,0 +1,196 @@
+package gitops
+
+import (
+	"context"
+	"os"
+	"path/filepath"
+
+	"github.com/pkg/errors"
+	"github.com/spf13/afero"
+	"gopkg.in/src-d/go-git.v4"
+	"gopkg.in/src-d/go-git.v4/plumbing"
+	"gopkg.in/src-d/go-git.v4/plumbing/transport"
+	"sigs.k8s.io/kustomize/pkg/gvk"
+	"sigs.k8s.io/kustomize/pkg/patch"
+	"sigs.k8s.io/kustomize/pkg/types"
+	"sigs.k8s.io/yaml"
+)
+
+const (
+	overlaysDir = "overlays/cluster-components"
+	baseDir     = "base"
+
+	defaultProfileGitHost = "github.com"
+	defaultProfileGitPath = "weaveworks/eks-gitops-example.git"
+)
+
+// GitOptions holds options for cloning a git repository
+type GitOptions struct {
+	URL    string
+	Branch string
+}
+
+// CloneFunc is a function that clones a Git repository
+type CloneFunc func(ctx context.Context, path string, options *git.CloneOptions) error
+
+// Profile represents a GitOps profile
+type Profile struct {
+	ClusterName string
+	Path        string
+	CloneFunc   CloneFunc
+	Fs          afero.Fs
```

Comment:
Nice with the dependency injection! Perhaps we can encapsulate the Io operations into a package with an interface so firstly we share that code and we always use the same libraries and secondly, we then mock our own code instead of a third party library.

Reply:
afero also provides a memory-backed filesystem that I have used in my tests. If we don't use a library like afero, we'd have to rewrite the same interfaces and a wrapper implementation that delegates to filesystem operations provided by the standard library. The mock code would also require checking if write operations were called with the right arguments.

Reply:
Ah! But then if we have the possibility to switch to in-memory calls with no side effects for the tests we don't need to encapsulate it. My concern was having to mock it during tests.

Reply:
For tests, I'm injecting the memory-based filesystem, so no mocking is required.